File: sdk-js/src/validationClient.ts
import { TollaraHeaders } from './tollaraHeaders';
import { DEFAULT_API_URL } from './constants';
import { resolveCorePathPrefix } from './pathPrefixes';
import { calculateHmac, constantTimeEquals, validateHmacSignature } from './hmac';
import { joinUrl, resolveBaseUrl } from './urls';
import { grantAccess } from './grantAccess';
import { parseUsageBreakdown, type UsageBreakdown } from './usageBreakdown';

export { grantAccess };

export interface ServiceKeyValidationResult {
  userId: string | null;
  serviceId: string | null;
  /** Core service-key row id when present in the validate JSON body. */
  serviceKeyId: string | null;
  serviceProductId: string | null;
  roles: string[];
  subscriptionStatus: string | null;
  validationSchemaVersion: number;
  billingModelType: string | null;
  measurementType: string | null;
  unitLabel: string | null;
  /** Whether {@link subscriptionStatus} grants invoke access. */
  grantAccess: boolean;
}

export interface UsageEstimateResult {
  sufficientCredits: boolean;
  wouldExceedCap: boolean;
  wouldAllow: boolean;
  estimatedCost: number | null;
  billingModelType: string | null;
  measurementType: string | null;
  unitLabel: string | null;
  breakdown: UsageBreakdown | null;
  estimateSchemaVersion: number;
  timestamp: number;
  httpStatus: number;
}

/** Canonical failure codes for validateServiceKeyWithOutcome (see docs-sdk §2.1.1). */
export type ValidationFailureCode =
  | 'MISSING_KEY'
  | 'NETWORK'
  | 'HTTP_ERROR'
  | 'MISSING_SIGNATURE_HEADERS'
  | 'HMAC_MISMATCH'
  | 'INVALID_KEY'
  | 'PARSE_ERROR';

export type ServiceKeyValidationOutcome =
  | { ok: true; result: ServiceKeyValidationResult }
  | { ok: false; code: ValidationFailureCode; message?: string; httpStatus?: number };

interface ValidateResponseBody {
  valid?: boolean;
  serviceKeyId?: string;
  userId?: string;
  serviceId?: string;
  serviceProductId?: string;
  roles?: string[];
  subscriptionStatus?: string;
  validationSchemaVersion?: number;
  billingModelType?: string | null;
  measurementType?: string | null;
  unitLabel?: string | null;
  error?: string;
}

/** Unsigned 401/403 from Core: `{ valid: false, error?: string }`. */
function invalidKeyFromUnsignedErrorBody(
  responseText: string,
  httpStatus: number,
): ServiceKeyValidationOutcome | null {
  if (httpStatus !== 401 && httpStatus !== 403) {
    return null;
  }
  try {
    const data = JSON.parse(responseText) as ValidateResponseBody;
    if (data.valid === false) {
      return {
        ok: false,
        code: 'INVALID_KEY',
        message: typeof data.error === 'string' ? data.error : undefined,
        httpStatus,
      };
    }
  } catch {
    // not JSON
  }
  return null;
}

function parseValidationResult(
  data: ValidateResponseBody,
  serviceId: string | null,
): ServiceKeyValidationResult {
  const subscriptionStatus =
    typeof data.subscriptionStatus === 'string' ? data.subscriptionStatus : null;

  return {
    userId: data.userId ?? null,
    serviceId: data.serviceId ?? serviceId ?? null,
    serviceKeyId:
      typeof data.serviceKeyId === 'string' && data.serviceKeyId.length > 0 ? data.serviceKeyId : null,
    serviceProductId: typeof data.serviceProductId === 'string' ? data.serviceProductId : null,
    roles: Array.isArray(data.roles) ? data.roles : [],
    subscriptionStatus,
    validationSchemaVersion:
      typeof data.validationSchemaVersion === 'number' ? data.validationSchemaVersion : 0,
    billingModelType: data.billingModelType ?? null,
    measurementType: data.measurementType ?? null,
    unitLabel: data.unitLabel ?? null,
    grantAccess: grantAccess(subscriptionStatus),
  };
}

/**
 * Validates a service key via the Tollara API and verifies response HMAC.
 * Returns a discriminated outcome with canonical failure codes (§2.1.1).
 */
export async function validateServiceKeyWithOutcome(params: {
  baseUrl?: string | null;
  serviceKey: string;
  serviceId: string | null;
  serviceSecret: string;
  fetch?: typeof globalThis.fetch;
}): Promise<ServiceKeyValidationOutcome> {
  const { baseUrl, serviceKey, serviceId, serviceSecret, fetch: fetchFn = fetch } = params;
  if (!serviceKey?.trim()) {
    return { ok: false, code: 'MISSING_KEY' };
  }

  const origin = resolveBaseUrl(baseUrl, DEFAULT_API_URL);
  const url = `${joinUrl(origin, resolveCorePathPrefix(baseUrl))}/service-keys/validate`;
  const body = JSON.stringify({ serviceKey, serviceId, serviceSecret });

  let res: Response;
  try {
    res = await fetchFn(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body,
    });
  } catch {
    return { ok: false, code: 'NETWORK' };
  }

  const httpStatus = res.status;
  const responseText = await res.text();
  if (!res.ok) {
    const unsignedInvalid = invalidKeyFromUnsignedErrorBody(responseText, httpStatus);
    if (unsignedInvalid) {
      return unsignedInvalid;
    }
    return { ok: false, code: 'HTTP_ERROR', httpStatus };
  }

  const signature = res.headers.get(TollaraHeaders.SIGNATURE);
  const timestamp = res.headers.get(TollaraHeaders.TIMESTAMP);
  if (!signature || !timestamp) {
    return { ok: false, code: 'MISSING_SIGNATURE_HEADERS', httpStatus };
  }

  const dataToVerify = responseText + timestamp;
  const expectedSig = calculateHmac(dataToVerify, serviceSecret);
  if (!constantTimeEquals(expectedSig, signature)) {
    return { ok: false, code: 'HMAC_MISMATCH', httpStatus };
  }

  let data: ValidateResponseBody;
  try {
    data = JSON.parse(responseText) as ValidateResponseBody;
  } catch {
    return { ok: false, code: 'PARSE_ERROR', httpStatus };
  }

  if (!data.valid) {
    return {
      ok: false,
      code: 'INVALID_KEY',
      message: typeof data.error === 'string' ? data.error : undefined,
      httpStatus,
    };
  }

  return { ok: true, result: parseValidationResult(data, serviceId) };
}

/**
 * Validates a service key via the Tollara API and verifies response HMAC.
 */
export async function validateServiceKey(
  params: {
    baseUrl?: string | null;
    serviceKey: string;
    serviceId: string | null;
    serviceSecret: string;
    fetch?: typeof globalThis.fetch;
  }
): Promise<ServiceKeyValidationResult | null> {
  const outcome = await validateServiceKeyWithOutcome(params);
  return outcome.ok ? outcome.result : null;
}

function parseEstimateResult(data: Record<string, unknown>, httpStatus: number): UsageEstimateResult {
  return {
    sufficientCredits: Boolean(data.sufficientCredits),
    wouldExceedCap: Boolean(data.wouldExceedCap),
    wouldAllow: Boolean(data.wouldAllow),
    estimatedCost: typeof data.estimatedCost === 'number' ? data.estimatedCost : null,
    billingModelType: typeof data.billingModelType === 'string' ? data.billingModelType : null,
    measurementType: typeof data.measurementType === 'string' ? data.measurementType : null,
    unitLabel: typeof data.unitLabel === 'string' ? data.unitLabel : null,
    breakdown: parseUsageBreakdown(data.breakdown),
    estimateSchemaVersion: typeof data.estimateSchemaVersion === 'number' ? data.estimateSchemaVersion : 0,
    timestamp: typeof data.timestamp === 'number' ? data.timestamp : 0,
    httpStatus,
  };
}

/**
 * Usage pre-flight for a service key (Core). Same trust model as {@link validateServiceKey}.
 */
export async function estimateUsage(params: {
  baseUrl?: string | null;
  serviceKey: string;
  serviceId: string | null;
  serviceSecret: string;
  estimatedUnits: number;
  fetch?: typeof globalThis.fetch;
}): Promise<UsageEstimateResult | null> {
  const { baseUrl, serviceKey, serviceId, serviceSecret, estimatedUnits, fetch: fetchFn = fetch } = params;
  if (!serviceKey?.trim()) return null;
  if (estimatedUnits == null || !Number.isFinite(estimatedUnits) || estimatedUnits <= 0) return null;

  const origin = resolveBaseUrl(baseUrl, DEFAULT_API_URL);
  const url = `${joinUrl(origin, resolveCorePathPrefix(baseUrl))}/service-keys/estimate-usage`;
  const body = JSON.stringify({ serviceKey, serviceId, serviceSecret, estimatedUnits });

  let res: Response;
  try {
    res = await fetchFn(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body,
    });
  } catch {
    return null;
  }

  const code = res.status;
  if (code !== 200 && code !== 403 && code !== 429) return null;

  const responseText = await res.text();
  if (!responseText.trim()) return null;

  const signature = res.headers.get(TollaraHeaders.SIGNATURE);
  const timestamp = res.headers.get(TollaraHeaders.TIMESTAMP);
  if (!signature || !timestamp) return null;
  if (!validateHmacSignature(signature, responseText + timestamp, serviceSecret)) return null;

  let data: Record<string, unknown>;
  try {
    data = JSON.parse(responseText) as Record<string, unknown>;
  } catch {
    return null;
  }

  return parseEstimateResult(data, code);
}

/** Core JWT usage estimate (`POST …/billing/usage/estimate`). Not HMAC-signed (spec §2.2). */
export async function estimateUsageWithJwt(params: {
  baseUrl?: string | null;
  corePathPrefix?: string | null;
  bearerToken: string;
  userId: string;
  serviceId: string;
  estimatedUnits: number;
  fetch?: typeof globalThis.fetch;
}): Promise<UsageEstimateResult | null> {
  const {
    baseUrl,
    corePathPrefix,
    bearerToken,
    userId,
    serviceId,
    estimatedUnits,
    fetch: fetchFn = fetch,
  } = params;
  if (!bearerToken?.trim() || !userId?.trim() || !serviceId?.trim()) return null;
  if (estimatedUnits == null || !Number.isFinite(estimatedUnits) || estimatedUnits <= 0) return null;

  const origin = resolveBaseUrl(baseUrl, DEFAULT_API_URL);
  const prefix = resolveCorePathPrefix(baseUrl, corePathPrefix);
  const url = `${joinUrl(origin, prefix)}/billing/usage/estimate`;
  const body = JSON.stringify({ userId, serviceId, estimatedUnits });

  let res: Response;
  try {
    res = await fetchFn(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${bearerToken.trim()}`,
      },
      body,
    });
  } catch {
    return null;
  }

  const code = res.status;
  if (code !== 200 && code !== 403 && code !== 429) return null;
  const responseText = await res.text();
  if (!responseText.trim()) return null;

  let data: Record<string, unknown>;
  try {
    data = JSON.parse(responseText) as Record<string, unknown>;
  } catch {
    return null;
  }

  return parseEstimateResult(data, code);
}

const CACHE_TTL_MS = 60_000;

interface CachedResult {
  result: ServiceKeyValidationResult;
  ts: number;
}

/** Simple cache for validateServiceKey (optional). */
export function createValidationCache() {
  const cache = new Map<string, CachedResult>();
  return {
    get(serviceKey: string): ServiceKeyValidationResult | null {
      const entry = cache.get(serviceKey);
      if (!entry || Date.now() - entry.ts > CACHE_TTL_MS) return null;
      return entry.result;
    },
    set(serviceKey: string, result: ServiceKeyValidationResult) {
      cache.set(serviceKey, { result, ts: Date.now() });
    },
    clear() {
      cache.clear();
    },
  };
}

File: sdk-python/src/tollara_service_sdk/validation_client.py
from dataclasses import dataclass
from enum import Enum
import json
from typing import Any, Dict, List, Literal, Optional, Union
from uuid import UUID

from .tollara_headers import TollaraHeaders
from .hmac_utils import validate_hmac_signature
from .usage_breakdown import UsageBreakdown, parse_usage_breakdown
from .verifier import grant_access

DEFAULT_CORE_PATH_PREFIX = "/api/v1"


def _core_service_keys_url(base_url: str, core_path_prefix: Optional[str], suffix: str) -> str:
    base = base_url.rstrip("/")
    p = (core_path_prefix or DEFAULT_CORE_PATH_PREFIX).strip()
    if not p.startswith("/"):
        p = "/" + p
    p = p.rstrip("/")
    return f"{base}{p}/service-keys/{suffix}"


def _validate_url(base_url: str, core_path_prefix: Optional[str]) -> str:
    return _core_service_keys_url(base_url, core_path_prefix, "validate")


def _optional_uuid(value: Any) -> Optional[UUID]:
    if value is None:
        return None
    if isinstance(value, UUID):
        return value
    s = str(value).strip()
    if not s:
        return None
    try:
        return UUID(s)
    except ValueError:
        return None


@dataclass
class ServiceKeyValidationResult:
    user_id: Optional[str]
    service_id: Optional[str]
    service_key_id: Optional[UUID]
    service_product_id: Optional[str]
    roles: List[str]
    subscription_status: Optional[str]
    validation_schema_version: int
    billing_model_type: Optional[str]
    measurement_type: Optional[str]
    unit_label: Optional[str]

    def grant_access(self) -> bool:
        return grant_access(self.subscription_status)

    @staticmethod
    def grant_access_for_status(subscription_status: Optional[str]) -> bool:
        return grant_access(subscription_status)


@dataclass
class UsageEstimateResult:
    sufficient_credits: bool
    would_exceed_cap: bool
    would_allow: bool
    estimated_cost: Optional[float]
    billing_model_type: Optional[str]
    measurement_type: Optional[str]
    unit_label: Optional[str]
    breakdown: Optional[UsageBreakdown]
    estimate_schema_version: int
    timestamp: int
    http_status: int


class ValidationFailureCode(str, Enum):
    MISSING_KEY = "MISSING_KEY"
    NETWORK = "NETWORK"
    HTTP_ERROR = "HTTP_ERROR"
    MISSING_SIGNATURE_HEADERS = "MISSING_SIGNATURE_HEADERS"
    HMAC_MISMATCH = "HMAC_MISMATCH"
    INVALID_KEY = "INVALID_KEY"
    PARSE_ERROR = "PARSE_ERROR"


@dataclass
class ServiceKeyValidationFailure:
    code: ValidationFailureCode
    message: Optional[str] = None
    http_status: Optional[int] = None


ServiceKeyValidationOutcome = Union[ServiceKeyValidationResult, ServiceKeyValidationFailure]


def _parse_validation_result(data: Dict[str, Any], service_id: Optional[str]) -> ServiceKeyValidationResult:
    roles = data.get("roles") or []
    return ServiceKeyValidationResult(
        user_id=data.get("userId"),
        service_id=data.get("serviceId") or service_id,
        service_key_id=_optional_uuid(data.get("serviceKeyId")),
        service_product_id=data.get("serviceProductId"),
        roles=roles if isinstance(roles, list) else [],
        subscription_status=data.get("subscriptionStatus"),
        validation_schema_version=int(data.get("validationSchemaVersion", 0)),
        billing_model_type=data.get("billingModelType"),
        measurement_type=data.get("measurementType"),
        unit_label=data.get("unitLabel"),
    )


def _invalid_key_from_unsigned_error_body(
    response_text: str, http_status: int
) -> Optional[ServiceKeyValidationFailure]:
    """Unsigned 401/403 from Core: ``{ \"valid\": false, \"error\"?: string }``."""
    if http_status not in (401, 403):
        return None
    try:
        data = json.loads(response_text)
    except ValueError:
        return None
    if isinstance(data, dict) and data.get("valid") is False:
        err = data.get("error")
        return ServiceKeyValidationFailure(
            code=ValidationFailureCode.INVALID_KEY,
            message=str(err) if err is not None else None,
            http_status=http_status,
        )
    return None


def validate_service_key_with_outcome(
    base_url: str,
    service_key: str,
    service_secret: str,
    service_id: Optional[str] = None,
    *,
    core_path_prefix: Optional[str] = None,
    session: Optional["requests.Session"] = None,
) -> ServiceKeyValidationOutcome:
    """Validate service key via Core service; returns result or structured failure (§2.1.1)."""
    if not service_key or not str(service_key).strip():
        return ServiceKeyValidationFailure(code=ValidationFailureCode.MISSING_KEY)
    try:
        import requests
    except ImportError:
        raise ImportError("validate_service_key_with_outcome requires 'requests'. pip install requests")
    url = _validate_url(base_url, core_path_prefix)
    body = {"serviceKey": service_key, "serviceId": service_id, "serviceSecret": service_secret}
    sess = session or requests.Session()
    try:
        resp = sess.post(url, json=body)
    except requests.RequestException:
        return ServiceKeyValidationFailure(code=ValidationFailureCode.NETWORK)
    http_status = resp.status_code
    response_text = resp.text
    if not resp.ok:
        unsigned_invalid = _invalid_key_from_unsigned_error_body(response_text, http_status)
        if unsigned_invalid is not None:
            return unsigned_invalid
        return ServiceKeyValidationFailure(code=ValidationFailureCode.HTTP_ERROR, http_status=http_status)
    signature = resp.headers.get(TollaraHeaders.SIGNATURE)
    timestamp = resp.headers.get(TollaraHeaders.TIMESTAMP)
    if not signature or not timestamp:
        return ServiceKeyValidationFailure(
            code=ValidationFailureCode.MISSING_SIGNATURE_HEADERS, http_status=http_status
        )
    if not validate_hmac_signature(signature, response_text + timestamp, service_secret):
        return ServiceKeyValidationFailure(code=ValidationFailureCode.HMAC_MISMATCH, http_status=http_status)
    try:
        data = json.loads(response_text)
    except ValueError:
        return ServiceKeyValidationFailure(code=ValidationFailureCode.PARSE_ERROR, http_status=http_status)
    if data.get("valid") is False:
        err = data.get("error")
        return ServiceKeyValidationFailure(
            code=ValidationFailureCode.INVALID_KEY,
            message=str(err) if err is not None else None,
            http_status=http_status,
        )
    return _parse_validation_result(data, service_id)


def validate_service_key(
    base_url: str,
    service_key: str,
    service_secret: str,
    service_id: Optional[str] = None,
    *,
    core_path_prefix: Optional[str] = None,
    session: Optional["requests.Session"] = None,
) -> Optional[ServiceKeyValidationResult]:
    """Validate service key via Core service. Requires 'requests' (pip install requests)."""
    outcome = validate_service_key_with_outcome(
        base_url, service_key, service_secret, service_id,
        core_path_prefix=core_path_prefix, session=session,
    )
    if isinstance(outcome, ServiceKeyValidationFailure):
        return None
    return outcome


def estimate_usage(
    base_url: str,
    service_key: str,
    service_secret: str,
    estimated_units: Union[int, float],
    service_id: Optional[str] = None,
    *,
    core_path_prefix: Optional[str] = None,
    session: Optional["requests.Session"] = None,
) -> Optional[UsageEstimateResult]:
    """Usage pre-flight via Core (same trust model as :func:`validate_service_key`). Requires ``requests``."""
    try:
        import requests
    except ImportError:
        raise ImportError("estimate_usage requires 'requests'. pip install requests")
    if estimated_units is None or float(estimated_units) <= 0:
        return None
    if not service_key or not str(service_key).strip():
        return None
    url = _core_service_keys_url(base_url, core_path_prefix, "estimate-usage")
    body = {
        "serviceKey": service_key,
        "serviceId": service_id,
        "serviceSecret": service_secret,
        "estimatedUnits": estimated_units,
    }
    sess = session or requests.Session()
    resp = sess.post(url, json=body)
    code = resp.status_code
    if code not in (200, 403, 429):
        return None
    response_text = resp.text
    if not response_text.strip():
        return None
    signature = resp.headers.get(TollaraHeaders.SIGNATURE)
    timestamp = resp.headers.get(TollaraHeaders.TIMESTAMP)
    if not signature or not timestamp:
        return None
    if not validate_hmac_signature(signature, response_text + timestamp, service_secret):
        return None
    data = resp.json()
    breakdown_raw = data.get("breakdown")
    breakdown = parse_usage_breakdown(breakdown_raw) if isinstance(breakdown_raw, dict) else None
    return UsageEstimateResult(
        sufficient_credits=bool(data.get("sufficientCredits")),
        would_exceed_cap=bool(data.get("wouldExceedCap")),
        would_allow=bool(data.get("wouldAllow")),
        estimated_cost=data.get("estimatedCost"),
        billing_model_type=data.get("billingModelType"),
        measurement_type=data.get("measurementType"),
        unit_label=data.get("unitLabel"),
        breakdown=breakdown,
        estimate_schema_version=int(data.get("estimateSchemaVersion", 0)),
        timestamp=int(data.get("timestamp", 0)),
        http_status=code,
    )
